File: programming/backtrack/46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python
#
# [46] 全排列
#

# @lc code=start
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def dfs(nums, size, path, used, res):
            if len(path) == size:
                res.append(path[:]) # path是引用，要进行值拷贝
                return
            # 查找临近点
            for i in range(size):
                # 判断条件
                if not used[i]:
                    used[i] = True
                    path.append(nums[i])

                    dfs(nums, size, path, used, res)

                    used[i] = False
                    path.pop()

        size = len(nums)
        if len(nums) == 0:
            return []

        used = [False for _ in range(size)]
        res = []
        dfs(nums, size, [], used, res)
        return res
        
# @lc code=end

File: programming/backtrack/46.全排列.cpp
class Solution {
public:
    // 需要用一个标记vis

    void backtrace(vector<int>& nums, vector<vector<int>>& res, vector<int>& tmp, 
                   vector<int>& vis) {
        if(tmp.size() == nums.size()) {
            res.push_back(tmp);
            return;
        }
        // 每次都从0开始，使用标记。。。。。
        for(int i=0; i < nums.size(); i++) {
            if(vis[i] == 0) {
                vis[i] = 1;
                tmp.push_back(nums[i]);
                backtrace(nums, res, tmp, vis);
                tmp.pop_back();
                vis[i] = 0;
            }

        }
        return;

    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> tmp;
        vector<int> vis(nums.size(), 0);
        if(nums.empty()) return res;
        int size = nums.size();
        backtrace(nums, res, tmp, vis);
        return res;
    }
};